bigtable/client/internal/endian.h

```diff
@@ -0,0 +1,150 @@
+
+#include <bitset>
+#include <limits>
+#include <vector>
+
+#include "bigtable/client/internal/strong_type.h"
+#include "bigtable/client/internal/throw_delegate.h"
+
+#ifdef _MSC_VER
+#include <stdlib.h>
+#elif defined(__GNUC__) || defined(__clang__)
+#include <byteswap.h>
+#endif
+
+#include <string>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+using bigendian64_t = internal::StrongType<std::int64_t, struct BigEndianType>;
+
+namespace internal {
+
+#ifndef _MSC_VER
+#ifndef __GNUC__
+#ifndef __clang__
+static std::int64_t explicit_swap64(std::int64_t value) {
+  std::int64_t return_value = ((((value)&0xff00000000000000ull) >> 56) |
+                               (((value)&0x00ff000000000000ull) >> 40) |
+                               (((value)&0x0000ff0000000000ull) >> 24) |
+                               (((value)&0x000000ff00000000ull) >> 8) |
+                               (((value)&0x00000000ff000000ull) << 8) |
+                               (((value)&0x0000000000ff0000ull) << 24) |
+                               (((value)&0x000000000000ff00ull) << 40) |
+                               (((value)&0x00000000000000ffull) << 56));
+
+  return return_value;
+}
+#endif
+#endif
+#endif
+
+inline bigtable::bigendian64_t byteswap64(bigtable::bigendian64_t value) {
+#ifdef _MSC_VER
+  return bigtable::bigendian64_t(_byteswap_uint64(value.get()));
+#elif defined(__GNUC__) || defined(__clang__)
+  return bigtable::bigendian64_t(__builtin_bswap64(value.get()));
+#else
+  return bigtable::bigendian64_t(explicit_swap64(value.get()));
+#endif
+}
+
+constexpr char c[sizeof(int)] = {1};
+constexpr bool IsBigEndian() {
+  return c[0] != 1;  // ignore different type comparison
+}
+
+template <typename T>
+struct encoder {
+  static std::string encode(T const& value);
+  static T decode(std::string const& value);
+};
+
+template <>
+struct encoder<bigtable::bigendian64_t> {
+  /**
+   * Convert BigEndian numeric value into a string of bytes and return it.
+   *
+   * Google Cloud Bigtable stores arbitrary blobs in each cell. These
+   * blobs are stored in a cell as a string of bytes. Values need to be
+   * converted to/from these cell blobs bytes to be used in application.
+   * This function is used to convert BigEndian 64 bit numeric value into
+   * string of bytes, so that it could be stored as a cell blob.
+   * For this conversion we consider the char size is 8 bits.
+   * So If machine architecture does not define char as 8 bit then we
+   * raise the assert.
+   *
+   * @param value reference to a number need to be converted into
+   *        string of bytes.
+   */
+  static std::string encode(bigtable::bigendian64_t const& value) {
+    static_assert(std::numeric_limits<unsigned char>::digits == 8,
+                  "This code assumes char is an 8-bit number");
+
+    std::string big_endian_string("", 8);
+    if (IsBigEndian()) {
+      std::memcpy((void*)big_endian_string.data(), &value, sizeof(value));
```

Comment:
Is the explicit cast to `void*` needed?  If so, please use `static_cast<void*>(...)`, if not, just remove it.